File: 1115. Print FooBar Alternately/1115. Print FooBar Alternately.cpp
class FooBar {
private:
    int n;
    atomic<bool> is_foo = true;
public:
    FooBar(int n) {
        this->n = n;
    }

    void foo(function<void()> printFoo) {
        for (int i = 0; i < n; i++) {
            while (!is_foo) {
                this_thread::yield();
            }
            
        	// printFoo() outputs "foo". Do not change or remove this line.
        	printFoo();
            
            is_foo = false;
        }
    }

    void bar(function<void()> printBar) {
        for (int i = 0; i < n; i++) {
            while (is_foo) {
                this_thread::yield();
            }
            
        	// printBar() outputs "bar". Do not change or remove this line.
        	printBar();
            
            is_foo = true;
        }
    }
};

File: 1115. Print FooBar Alternately/1115. Print FooBar Alternately.rs
struct FooBar {
    n: usize,
    is_foo: Arc<Mutex<bool>>,
    condvar: Arc<Condvar>,
}

impl FooBar {
    fn new(n: usize) -> Self {
        FooBar {
            n,
            is_foo: Arc::new(Mutex::new(true)),
            condvar: Arc::new(Condvar::new()),
        }
    }

    fn foo<F>(&self, print_foo: F)
    where
        F: Fn(),
    {
        for _ in 0..self.n {
            let mut is_foo_val = self.condvar.wait_while(self.is_foo.lock().unwrap(), |b| { !*b }).unwrap();
            // printFoo() outputs "foo". Do not change or remove this line.
            print_foo();
            *is_foo_val = false;
            self.condvar.notify_one();
        }
    }

    fn bar<F>(&self, print_bar: F)
    where
        F: Fn(),
    {
        for _ in 0..self.n {
            let mut is_foo_val = self.condvar.wait_while(self.is_foo.lock().unwrap(), |b| { *b }).unwrap();
            // printBar() outputs "bar". Do not change or remove this line.
            print_bar();
            *is_foo_val = true;
            self.condvar.notify_one();
        }
    }
}
